File: weird_captcha_gym/shared_runtime/app/mechanics/ballast_lantern.js
(() => {
  "use strict";

  const MECHANIC_ID = "ballast_lantern";
  const TRACK_UNITS = 10000;
  let activeCleanup = null;

  const clone = (value) => JSON.parse(JSON.stringify(value));
  const truncDiv = (numerator, denominator) => numerator >= 0 ? Math.floor(numerator / denominator) : -Math.floor(-numerator / denominator);

  function advanceSpecimen(sim, motion) {
    const tick = sim.tick;
    const law = motion.law;
    let velocity = sim.specimen_velocity;
    if (law === "darter" && (tick - 1) % motion.darter_interval === 0) {
      const index = Math.floor((tick - 1) / motion.darter_interval) % motion.darter_velocities.length;
      velocity = motion.darter_velocities[index];
    } else if (law === "steady_sinker") {
      velocity = Math.max(-motion.speed, velocity - motion.acceleration);
    } else if (law === "floater") {
      velocity = Math.min(motion.speed, velocity + motion.acceleration);
    }
    let position = sim.specimen_y + velocity;
    if (position <= motion.min_y) {
      position = motion.min_y;
      velocity = law === "steady_sinker" ? motion.boundary_burst : Math.abs(velocity);
    } else if (position >= motion.max_y) {
      position = motion.max_y;
      velocity = law === "floater" ? -motion.boundary_burst : -Math.abs(velocity);
    }
    sim.specimen_y = position;
    sim.specimen_velocity = velocity;
  }

  function advance(model) {
    const {sim, state, engaged} = model;
    const p = state.parameters;
    const motion = state.motion;
    const crate = state.crate;
    if (sim.status !== "active") return;
    sim.tick += 1;
    let velocity = sim.cage_velocity + (engaged ? p.thrust_accel : -p.gravity_accel);
    velocity = truncDiv(velocity * p.drag_numerator, p.drag_denominator);
    let position = sim.cage_y + velocity;
    const lower = p.cage_half_height;
    const upper = TRACK_UNITS - p.cage_half_height;
    if (position <= lower) {
      position = lower;
      velocity = truncDiv(Math.abs(velocity) * p.boundary_restitution_numerator, p.boundary_restitution_denominator);
    } else if (position >= upper) {
      position = upper;
      velocity = -truncDiv(Math.abs(velocity) * p.boundary_restitution_numerator, p.boundary_restitution_denominator);
    }
    sim.cage_y = position;
    sim.cage_velocity = velocity;
    advanceSpecimen(sim, motion);
    sim.crate_spawned = sim.tick >= state.crate.spawn_tick;
    sim.specimen_inside = Math.abs(sim.cage_y - sim.specimen_y) <= p.cage_half_height - p.specimen_half_height;
    sim.crate_inside = Boolean(sim.crate_spawned && Math.abs(sim.cage_y - state.crate.y) <= p.cage_half_height - p.crate_half_height);
    sim.capture_meter = sim.specimen_inside
      ? Math.min(p.capture_max, sim.capture_meter + p.capture_fill_per_tick)
      : Math.max(0, sim.capture_meter - p.capture_drain_per_tick);
    if (sim.crate_inside) sim.crate_meter = Math.min(p.crate_meter_max, sim.crate_meter + p.crate_fill_per_tick);
    if (sim.capture_meter <= 0) sim.status = "escaped";
    else if (sim.capture_meter >= p.capture_max) sim.status = sim.crate_meter >= p.crate_meter_max ? "secured" : "specimen_only";
    else if (sim.tick >= p.max_ticks) sim.status = "timeout";
  }

  function recordWinch(model, engaged, inputSource, phase) {
    tick(model);
    if (model.sim.status !== "active" || model.submitting || model.engaged === engaged) return;
    const action = model.helpers.beginAction?.("ballast-winch-transition");
    model.engaged = engaged;
    model.events.push({
      sequence: model.events.length + 1,
      type: "winch",
      tick: model.sim.tick,
      engaged,
      input_source: inputSource,
      phase,
    });
    model.helpers.setReadout(engaged ? "HAUL" : "COAST", "idle");
    updateControls(model);
    action?.settle();
  }

  function snapshot(sim) {
    return Object.fromEntries([
      "tick", "cage_y", "cage_velocity", "specimen_y", "specimen_velocity", "capture_meter",
      "crate_meter", "crate_spawned", "specimen_inside", "crate_inside", "status",
    ].map((key) => [key, sim[key]]));
  }

  function payload(model) {
    return {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      interaction_mode: model.interaction,
      events: clone(model.events),
      terminal_tick: model.sim.tick,
      final_state: snapshot(model.sim),
      completed: model.sim.status === "secured",
    };
  }

  function setVerdict(model, kind, title) {
    const root = document.querySelector(".ballast-lantern");
    const node = document.querySelector(".ballast-verdict");
    if (!root || !node) return;
    root.classList.toggle("is-passed", kind === "pass");
    root.classList.toggle("is-failed", kind === "fail");
    node.innerHTML = `<b>${model.helpers.text(title)}</b>`;
  }

  async function postPayload(model, outgoing) {
    if (model.submitting) return;
    model.submitting = true;
    model.pendingPayload = outgoing;
    updateControls(model);
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify(outgoing),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.completed = true;
        setVerdict(model, "pass", "PASS");
        model.helpers.setReadout("PASS", "passed");
        updateControls(model);
        return;
      }
      if (outcome.state) {
        await render(outcome.state, model.helpers, {freshFailure: true});
        const fresh = window.ballastLanternModel;
        setVerdict(fresh, "fail", "FAIL");
        fresh.helpers.setReadout("FAIL", "error");
      } else {
        model.submitting = false;
        model.helpers.setReadout("FAIL", "error");
        document.querySelector(".ballast-retry")?.removeAttribute("hidden");
        updateControls(model);
      }
    } catch (_error) {
      model.submitting = false;
      model.helpers.setReadout("FAIL", "error");
      document.querySelector(".ballast-retry")?.removeAttribute("hidden");
      updateControls(model);
    }
  }

  function terminal(model) {
    if (model.terminal) return;
    model.terminal = true;
    if (model.interval) window.cancelAnimationFrame(model.interval);
    model.interval = null;
    const title = model.sim.status === "secured" ? "PASS" : "FAIL";
    setVerdict(model, model.sim.status === "secured" ? "pass" : "fail", title);
    model.helpers.setReadout(title, model.sim.status === "secured" ? "passed" : "error");
    void postPayload(model, payload(model));
  }

  function pct(value) { return `${(Number(value) / TRACK_UNITS * 100).toFixed(3)}%`; }

  function updateControls(model) {
    const root = document.querySelector(".ballast-lantern");
    if (!root) return;
    root.dataset.engaged = String(model.engaged);
    root.dataset.terminal = String(model.terminal);
    root.dataset.interaction = model.interaction;
    document.querySelectorAll(".ballast-controls button").forEach((button) => { button.disabled = model.terminal || model.submitting; });
    const haul = document.querySelector(".ballast-haul");
    const coast = document.querySelector(".ballast-coast");
    if (haul) haul.dataset.active = String(model.engaged);
    if (coast) coast.dataset.active = String(!model.engaged);
  }

  function draw(model) {
    const {sim, state} = model;
    const p = state.parameters;
    const cage = document.querySelector(".ballast-cage");
    const specimen = document.querySelector(".ballast-specimen");
    const crate = document.querySelector(".ballast-crate");
    const cable = document.querySelector(".ballast-cable");
    if (!cage || !specimen || !crate || !cable) return;
    cage.style.bottom = pct(sim.cage_y);
    cage.style.height = pct(p.cage_half_height * 2);
    cage.dataset.specimen = String(sim.specimen_inside);
    cage.dataset.crate = String(sim.crate_inside);
    specimen.style.bottom = pct(sim.specimen_y);
    specimen.style.height = pct(p.specimen_half_height * 2);
    specimen.dataset.inside = String(sim.specimen_inside);
    crate.style.bottom = pct(state.crate.y);
    crate.style.height = pct(p.crate_half_height * 2);
    crate.dataset.spawned = String(sim.crate_spawned);
    crate.dataset.inside = String(sim.crate_inside);
    cable.style.height = pct(TRACK_UNITS - sim.cage_y);
    const trend = document.querySelector(".ballast-trend");
    if (trend) trend.dataset.direction = sim.specimen_velocity >= 0 ? "up" : "down";
    const reel = document.querySelector(".ballast-reel");
    reel.style.transform = `rotate(${(sim.tick * (model.engaged ? 9 : -5)) % 360}deg)`;
    document.querySelectorAll(".ballast-wake-dot").forEach((dot, index) => {
      dot.style.bottom = pct(model.trail[index] ?? sim.specimen_y);
      dot.style.opacity = String(Math.max(.08, .5 - index * .08));
    });
    updateControls(model);
  }

  function tick(model) {
    if (model.terminal || model.submitting) return;
    const target = Math.floor((performance.now() - model.started) / Number(model.state.parameters.tick_ms) + 1e-7);
    while (model.sim.tick < target && model.sim.status === "active") {
      advance(model);
      model.trail.unshift(model.sim.specimen_y);
      model.trail = model.trail.slice(0, model.state.parameters.trail_samples);
    }
    draw(model);
    if (model.sim.status !== "active") terminal(model);
  }

  async function abandon(model) {
    if (model.submitting || model.completed) return;
    if (model.interval) window.cancelAnimationFrame(model.interval);
    model.interval = null;
    model.terminal = true;
    model.helpers.setReadout("…", "pending");
    await postPayload(model, {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      interaction_mode: model.interaction,
      events: [],
      terminal_tick: 0,
      final_state: snapshot(model.sim),
      completed: false,
    });
  }

  async function render(state, helpers, options = {}) {
    if (activeCleanup) activeCleanup();
    document.body.dataset.mechanic = "ballast-lantern";
    const interaction = state.control_condition?.interaction || "full";
    const trailCount = Number(state.parameters.trail_samples);
    const model = {
      state,
      helpers,
      interaction,
      events: [],
      sim: clone(state.initial_state),
      engaged: false,
      terminal: false,
      completed: false,
      submitting: false,
      pendingPayload: null,
      interval: null,
      started: performance.now(),
      trail: Array.from({length: trailCount}, () => state.initial_state.specimen_y),
      keydown: null,
      keyup: null,
      blur: null,
    };
    window.ballastLanternModel = model;
    const wake = Array.from({length: trailCount}, (_, index) => `<i class="ballast-wake-dot" data-wake="${index}"></i>`).join("");
    const controls = interaction === "simplified"
      ? `<div class="ballast-controls"><button type="button" class="ballast-haul">HAUL</button><button type="button" class="ballast-coast" data-active="true">COAST</button></div>`
      : `<div class="ballast-spacebar"><b>SPACE</b></div>`;
    helpers.app.innerHTML = `
      <section class="ballast-lantern" data-engaged="false" data-terminal="false" data-interaction="${helpers.text(interaction)}" data-challenge-id="${helpers.text(state.challenge_id)}">
        <div class="ballast-verdict" aria-live="assertive"></div>
        <header class="ballast-head">
          <div class="ballast-mark"><i></i><span>BL<br>17</span></div>
          <div><small>PELAGIC RECOVERY OFFICE / NIGHT SHIFT</small><h1>Cage both signals. Ballast first.</h1></div>
        </header>
        <main class="ballast-workbench">
          <section class="ballast-shaft-shell">
            <div class="ballast-reel-house"><div class="ballast-reel"><i></i><b></b></div><span>ONE-CHANNEL WINCH</span></div>
            <div class="ballast-shaft" aria-label="Flooded vertical salvage shaft">
              <div class="ballast-depths"><span>00</span><span>25</span><span>50</span><span>75</span><span>100</span></div>
              <div class="ballast-cable"></div>
              ${wake}
              <div class="ballast-specimen" aria-label="moving luminous specimen"><i></i><b></b><span></span></div>
              <div class="ballast-crate" aria-label="ballast salvage crate"><i></i><b></b><span>Ⅱ</span></div>
              <div class="ballast-cage" aria-label="lantern capture cage"><i></i><b></b><span></span></div>
              <div class="ballast-waterline"></div>
            </div>
            ${state.parameters.show_trend_beacon ? `<div class="ballast-trend" data-direction="up" aria-label="specimen drift beacon"><i></i></div>` : ""}
          </section>
          <aside class="ballast-console">
            <div class="ballast-console-title"><small>WINCH TELEGRAPH</small><b>BL-17</b></div>
            <div class="ballast-dial"><div class="ballast-dial-hand"></div><strong>WINCH</strong></div>
            ${controls}
          </aside>
        </main>
        <footer class="ballast-foot"><div class="readout" data-status="idle">COAST</div><button type="button" class="ballast-retry" hidden>RETRY</button><button type="button" class="ballast-abandon">NEW SHAFT</button></footer>
        ${helpers.cheatPanelTemplate()}
      </section>`;

    model.keydown = (event) => {
      if (interaction !== "full" || event.code !== "Space" || event.repeat || model.terminal) return;
      event.preventDefault();
      recordWinch(model, true, "keyboard_hold", "keydown");
    };
    model.keyup = (event) => {
      if (interaction !== "full" || event.code !== "Space" || model.terminal) return;
      event.preventDefault();
      recordWinch(model, false, "keyboard_hold", "keyup");
    };
    model.blur = () => { if (interaction === "full" && model.engaged && !model.terminal) recordWinch(model, false, "keyboard_hold", "keyup"); };
    document.addEventListener("keydown", model.keydown);
    document.addEventListener("keyup", model.keyup);
    window.addEventListener("blur", model.blur);
    document.querySelector(".ballast-haul")?.addEventListener("click", () => recordWinch(model, true, "winch_button", "haul"));
    document.querySelector(".ballast-coast")?.addEventListener("click", () => recordWinch(model, false, "winch_button", "coast"));
    document.querySelector(".ballast-abandon")?.addEventListener("click", () => void abandon(model));
    document.querySelector(".ballast-retry")?.addEventListener("click", () => {
      document.querySelector(".ballast-retry")?.setAttribute("hidden", "");
      void postPayload(model, model.pendingPayload || payload(model));
    });
    if (helpers.isCheatMode()) helpers.installCheatPanel?.();
    draw(model);
    const frame = () => {
      tick(model);
      if (!model.terminal) model.interval = window.requestAnimationFrame(frame);
    };
    model.interval = window.requestAnimationFrame(frame);
    if (options.freshFailure) {
      window.setTimeout(() => document.querySelector(".ballast-lantern")?.classList.remove("is-failed"), 1450);
    }
    activeCleanup = () => {
      if (model.interval) window.cancelAnimationFrame(model.interval);
      document.removeEventListener("keydown", model.keydown);
      document.removeEventListener("keyup", model.keyup);
      window.removeEventListener("blur", model.blur);
      model.terminal = true;
    };
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics[MECHANIC_ID] = {rootSelector: ".ballast-lantern", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/ballast_lantern.py
from __future__ import annotations

from typing import Any


MECHANIC_ID = "ballast_lantern"
TRACK_UNITS = 10_000


def _fail(message: str, score: int = 0) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": score, "feedback": message}


def _integer(value: Any, label: str, lower: int | None = None, upper: int | None = None) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{label} is not an integer")
    if lower is not None and value < lower:
        raise ValueError(f"{label} is below its bound")
    if upper is not None and value > upper:
        raise ValueError(f"{label} is above its bound")
    return value


def _trunc_div(numerator: int, denominator: int) -> int:
    return numerator // denominator if numerator >= 0 else -((-numerator) // denominator)


def _identity(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> str | None:
    if any(str(value.get("mechanic_id") or "") != MECHANIC_ID for value in (payload, truth, public)):
        return "mechanic mismatch"
    for key in ("task_id", "challenge_id"):
        expected = str(truth.get(key) or "")
        if not expected or str(payload.get(key) or "") != expected or str(public.get(key) or "") != expected:
            return f"stale or mismatched {key}"
    return None


def _contract(truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    for key in ("track_units", "parameters", "motion", "crate", "initial_state"):
        if truth.get(key) != public.get(key):
            raise ValueError(f"public and hidden {key} disagree")
    if truth.get("control_condition") != public.get("control_condition"):
        raise ValueError("control condition disagrees")
    if truth.get("track_units") != TRACK_UNITS:
        raise ValueError("shaft scale is malformed")
    p = truth.get("parameters")
    motion = truth.get("motion")
    crate = truth.get("crate")
    initial = truth.get("initial_state")
    if not all(isinstance(value, dict) for value in (p, motion, crate, initial)):
        raise ValueError("world records are missing")
    condition = truth.get("control_condition")
    if condition is not None and condition.get("difficulty_parameters") != p:
        raise ValueError("condition parameters disagree")
    interaction = str((condition or {}).get("interaction") or "full")
    if interaction not in {"simplified", "full"}:
        raise ValueError("interaction mode is invalid")

    integer_bounds = {
        "tick_ms": (40, 100), "max_ticks": (500, 1500), "cage_half_height": (600, 1400),
        "specimen_half_height": (120, 260), "crate_half_height": (180, 400),
        "thrust_accel": (3, 8), "gravity_accel": (2, 7), "drag_numerator": (85, 99),
        "drag_denominator": (100, 100), "boundary_restitution_numerator": (0, 3),
        "boundary_restitution_denominator": (2, 5), "specimen_speed_min": (14, 52),
        "specimen_speed_max": (18, 64), "specimen_accel": (1, 4),
        "darter_interval_min": (14, 50), "darter_interval_max": (18, 60),
        "capture_initial": (2500, 5500), "capture_max": (8000, 12000),
        "capture_fill_per_tick": (16, 40), "capture_drain_per_tick": (5, 24),
        "crate_meter_max": (1800, 4500), "crate_fill_per_tick": (24, 64),
        "crate_spawn_tick_min": (60, 150), "crate_spawn_tick_max": (70, 170),
        "crate_min_separation": (600, 3500), "trail_samples": (1, 6),
    }
    for key, bounds in integer_bounds.items():
        _integer(p.get(key), key, *bounds)
    laws = p.get("motion_law_pool")
    if not isinstance(laws, list) or not laws or len(laws) != len(set(laws)) or not set(laws) <= {"steady_sinker", "darter", "floater", "oscillator"}:
        raise ValueError("motion-law pool is malformed")
    if not isinstance(p.get("show_trend_beacon"), bool):
        raise ValueError("trend-beacon condition is malformed")
    if p["capture_fill_per_tick"] <= p["capture_drain_per_tick"] or p["capture_initial"] >= p["capture_max"]:
        raise ValueError("capture meter condition is malformed")
    if p["cage_half_height"] <= max(p["specimen_half_height"], p["crate_half_height"]):
        raise ValueError("cage geometry is malformed")
    if p["specimen_speed_min"] > p["specimen_speed_max"] or p["darter_interval_min"] > p["darter_interval_max"]:
        raise ValueError("motion parameter range is inverted")
    if p["crate_spawn_tick_min"] > p["crate_spawn_tick_max"]:
        raise ValueError("crate spawn range is inverted")
    if p["boundary_restitution_numerator"] >= p["boundary_restitution_denominator"]:
        raise ValueError("shaft restitution must lose energy")

    law = str(motion.get("law") or "")
    if law not in laws:
        raise ValueError("generated drift law is outside the active pool")
    motion_keys = ("min_y", "max_y", "start_y", "start_velocity", "speed", "acceleration", "boundary_burst", "darter_interval")
    for key in motion_keys:
        _integer(motion.get(key), f"motion {key}")
    if not 0 < motion["min_y"] < motion["start_y"] < motion["max_y"] < TRACK_UNITS:
        raise ValueError("specimen track is malformed")
    velocities = motion.get("darter_velocities")
    if not isinstance(velocities, list) or len(velocities) != 40 or any(isinstance(value, bool) or not isinstance(value, int) for value in velocities):
        raise ValueError("darter schedule is malformed")
    spawn_tick = _integer(crate.get("spawn_tick"), "crate spawn", p["crate_spawn_tick_min"], p["crate_spawn_tick_max"])
    crate_y = _integer(crate.get("y"), "crate position", p["cage_half_height"], TRACK_UNITS - p["cage_half_height"])
    if spawn_tick >= p["max_ticks"] or not 0 < crate_y < TRACK_UNITS:
        raise ValueError("crate contract is malformed")
    expected_initial = {
        "tick": 0, "cage_y": max(p["cage_half_height"], 1850), "cage_velocity": 0,
        "specimen_y": motion["start_y"], "specimen_velocity": motion["start_velocity"],
        "capture_meter": p["capture_initial"], "crate_meter": 0, "crate_spawned": False,
        "specimen_inside": False, "crate_inside": False, "status": "active",
    }
    if initial != expected_initial:
        raise ValueError("initial simulation state is malformed")
    return {"parameters": p, "motion": motion, "crate": crate, "initial": initial, "interaction": interaction}


def _advance_specimen(sim: dict[str, Any], motion: dict[str, Any]) -> None:
    tick = sim["tick"]
    law = motion["law"]
    velocity = sim["specimen_velocity"]
    if law == "darter" and (tick - 1) % motion["darter_interval"] == 0:
        index = ((tick - 1) // motion["darter_interval"]) % len(motion["darter_velocities"])
        velocity = motion["darter_velocities"][index]
    elif law == "steady_sinker":
        velocity = max(-motion["speed"], velocity - motion["acceleration"])
    elif law == "floater":
        velocity = min(motion["speed"], velocity + motion["acceleration"])
    position = sim["specimen_y"] + velocity
    if position <= motion["min_y"]:
        position = motion["min_y"]
        velocity = motion["boundary_burst"] if law == "steady_sinker" else abs(velocity)
    elif position >= motion["max_y"]:
        position = motion["max_y"]
        velocity = -motion["boundary_burst"] if law == "floater" else -abs(velocity)
    sim["specimen_y"], sim["specimen_velocity"] = position, velocity


def _advance(sim: dict[str, Any], engaged: bool, p: dict[str, Any], motion: dict[str, Any], crate: dict[str, Any]) -> None:
    if sim["status"] != "active":
        return
    sim["tick"] += 1
    velocity = sim["cage_velocity"] + (p["thrust_accel"] if engaged else -p["gravity_accel"])
    velocity = _trunc_div(velocity * p["drag_numerator"], p["drag_denominator"])
    position = sim["cage_y"] + velocity
    lower, upper = p["cage_half_height"], TRACK_UNITS - p["cage_half_height"]
    if position <= lower:
        position = lower
        velocity = _trunc_div(abs(velocity) * p["boundary_restitution_numerator"], p["boundary_restitution_denominator"])
    elif position >= upper:
        position = upper
        velocity = -_trunc_div(abs(velocity) * p["boundary_restitution_numerator"], p["boundary_restitution_denominator"])
    sim["cage_y"], sim["cage_velocity"] = position, velocity
    _advance_specimen(sim, motion)
    sim["crate_spawned"] = sim["tick"] >= crate["spawn_tick"]
    sim["specimen_inside"] = abs(sim["cage_y"] - sim["specimen_y"]) <= p["cage_half_height"] - p["specimen_half_height"]
    sim["crate_inside"] = bool(sim["crate_spawned"] and abs(sim["cage_y"] - crate["y"]) <= p["cage_half_height"] - p["crate_half_height"])
    if sim["specimen_inside"]:
        sim["capture_meter"] = min(p["capture_max"], sim["capture_meter"] + p["capture_fill_per_tick"])
    else:
        sim["capture_meter"] = max(0, sim["capture_meter"] - p["capture_drain_per_tick"])
    if sim["crate_inside"]:
        sim["crate_meter"] = min(p["crate_meter_max"], sim["crate_meter"] + p["crate_fill_per_tick"])
    if sim["capture_meter"] <= 0:
        sim["status"] = "escaped"
    elif sim["capture_meter"] >= p["capture_max"]:
        sim["status"] = "secured" if sim["crate_meter"] >= p["crate_meter_max"] else "specimen_only"
    elif sim["tick"] >= p["max_ticks"]:
        sim["status"] = "timeout"


def _snapshot(sim: dict[str, Any]) -> dict[str, Any]:
    return {key: sim[key] for key in (
        "tick", "cage_y", "cage_velocity", "specimen_y", "specimen_velocity", "capture_meter",
        "crate_meter", "crate_spawned", "specimen_inside", "crate_inside", "status",
    )}


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    identity_error = _identity(payload, truth, public)
    if identity_error:
        return _fail(identity_error)
    try:
        contract = _contract(truth, public)
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"invalid Ballast Lantern contract: {exc}")
    interaction = contract["interaction"]
    if payload.get("interaction_mode") != interaction:
        return _fail("submitted interaction mode differs from task condition")
    events = payload.get("events")
    if not isinstance(events, list) or not events or len(events) > 10_000:
        return _fail("winch transcript is missing or oversized")
    terminal_tick = payload.get("terminal_tick")
    try:
        terminal_tick = _integer(terminal_tick, "terminal tick", 1, contract["parameters"]["max_ticks"])
    except ValueError as exc:
        return _fail(str(exc))
    previous_event_tick = -1
    for sequence, event in enumerate(events, 1):
        if not isinstance(event, dict) or event.get("sequence") != sequence or event.get("type") != "winch":
            return _fail(f"event {sequence} has invalid identity")
        event_tick = event.get("tick")
        if isinstance(event_tick, bool) or not isinstance(event_tick, int) or not previous_event_tick <= event_tick < terminal_tick:
            return _fail(f"event {sequence} has an invalid tick")
        previous_event_tick = event_tick
    expected_source = "keyboard_hold" if interaction == "full" else "winch_button"
    sim = dict(contract["initial"])
    engaged = False
    next_event = 0
    try:
        while sim["tick"] < terminal_tick and sim["status"] == "active":
            while next_event < len(events) and events[next_event].get("tick") == sim["tick"]:
                event = events[next_event]
                sequence = next_event + 1
                if event.get("input_source") != expected_source:
                    raise ValueError(f"event {sequence} uses the wrong interaction input")
                if not isinstance(event.get("engaged"), bool) or event["engaged"] == engaged:
                    raise ValueError(f"event {sequence} does not change the winch state")
                phase = event.get("phase")
                expected_phase = ("keydown" if event["engaged"] else "keyup") if interaction == "full" else ("haul" if event["engaged"] else "coast")
                if phase != expected_phase:
                    raise ValueError(f"event {sequence} reports the wrong control phase")
                engaged = event["engaged"]
                next_event += 1
            _advance(sim, engaged, contract["parameters"], contract["motion"], contract["crate"])
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"Ballast replay rejected: {exc}")
    if next_event != len(events):
        return _fail("winch transcript contains events after the submitted terminal state")
    if sim["tick"] != terminal_tick or sim["status"] == "active":
        return _fail("submitted run did not reach a terminal simulation state")
    if payload.get("final_state") != _snapshot(sim):
        return _fail("submitted final shaft state does not match replay")
    completed = sim["status"] == "secured"
    if payload.get("completed") is not completed:
        return _fail("submitted completion flag does not match replay")
    if completed:
        score = 100
    elif sim["status"] == "specimen_only":
        score = 65
    else:
        specimen = round(50 * sim["capture_meter"] / contract["parameters"]["capture_max"])
        crate = round(30 * sim["crate_meter"] / contract["parameters"]["crate_meter_max"])
        score = min(80, specimen + crate)
    return {
        "graded": True,
        "passed": completed,
        "score": score,
        "feedback": (
            f"replayed {sim['tick']} shaft ticks; specimen {sim['capture_meter']}/{contract['parameters']['capture_max']}; "
            f"crate {sim['crate_meter']}/{contract['parameters']['crate_meter_max']}; {sim['status']}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    del public_state
    return {
        "reference_schedule": ground_truth.get("reference_schedule"),
        "reference_final_state": ground_truth.get("reference_final_state"),
        "reference_metrics": ground_truth.get("reference_metrics"),
    }
